### PR TITLE
Support Sorbet-annotated task code display in task UI

### PR DESCRIPTION
**Pull request by tyetraskshopify**
Preamble: I'm not sure this is something that you all would like to support (especially since it adds Sorbet as a development/test dependency just to support the test). I figured I'd open the pull request and if it should be closed right away, that's fine from my side! 🎉 

Please let me know if I've missed any steps or if this is the wrong approach. If there are additional changes you'd like, I'm more than happy to make them as well. Thank you!

### Issue

Gem versions: `1.5.0`, `1.6.0`
Ruby versions: `2.7.2`, `3.0.2p107`
Rails versions: `6.0.4`, `6.1.4.1`

After creating a new task recently (and with `sorbet-runtime` enabled), I noticed that the source code display in the UI returns source for [Sorbet's `T::Private::Methods` module](https://github.com/sorbet/sorbet/blob/master/gems/sorbet-runtime/lib/types/private/methods/_methods.rb) instead of the maintenance task itself. This makes sense based on what I see in the current `MaintenanceTasks::TaskData#code` implementation. I realize that this might not be a common enough problem though depending on how often `sorbet-runtime` is enabled/disabled in production and further, how often developers are adding type annotations to tasks.

#### Expected (and fixed) behavior:

Task source code is visible when viewed in the UI.

Screen Shot 2021-11-09 at 10 36 11 AM

#### Current behavior:

Sorbet source code is visible when viewed in the UI.

Screen Shot 2021-11-09 at 10 36 36 AM

### Approach

[`Module::const_source_location` added in Ruby 2.7](https://ruby-doc.org/core-2.7.0/Module.html#method-i-const_source_location) appears to meet this need and returns the correct source whether using Sorbet or not. I've guarded against the method presence to allow backward compatibility with older Ruby versions. However, I see the target Ruby version might be 2.7+ so if I should remove the conditional, please let me know.

##### Fixes
- `MaintenanceTasks::Task`s with prepended modules (e.g. Sorbet-annotated tasks) will correctly display the source code of the task when viewed in the UI.

### CODE REVIEW

**Review by etiennebarrie — COMMENTED** [review]
> `const_source_location` was that it has some issues with autoloading

I think it should be ok since we have loaded the constant at that point?

---

While I understand the problem surfaced by using sorbet for you, can we avoid adding sorbet to our Gemfile, and reproduce the issue with a more minimal case. It could be a prepended module that defines `process`, even if that's not how sorbet works. I think we have a similar issue with statsd-instrument which does use prepend https://github.com/Shopify/statsd-instrument/#metaprogramming-methods

**Comment by tyetraskshopify** [comment]
> can we avoid adding sorbet to our Gemfile, and reproduce the issue with a more minimal case.

Absolutely! I'll work on that change and push it later this morning. Thank you!

**Comment by tyetraskshopify** [comment]
@nvasilevski, @etiennebarrie, thank you both! I've pushed the change using a simpler prepended module. Can you let me know if this is what you have in mind? If there are any other changes you'd like to see, I'm more than happy to make them!

Do either of you have an opinion on whether the conditional should remain in `MaintenanceTasks::TaskData#code` (or should it always use `Module::const_source_location`)? Also, please let me know if I should squash and re-push. I've left the individual commit for easier review for now. Thanks!

File: test/models/maintenance_tasks/task_data_test.rb [hunk 1]
```diff
@@ -50,6 +51,14 @@ class TaskDataTest < ActiveSupport::TestCase
         task_data.code.each_line.grep(/UpdatePostsTask/).first.squish
     end
 
+    test "#code returns the code source of a Task with a prepended module" do
+      task_klass = "UpdatePostsModulePrependedTask"
+      task_data = TaskData.new("Maintenance::#{task_klass}")
+
+      assert_equal "class #{task_klass} < MaintenanceTasks::Task",
+        task_data.code.each_line.grep(/#{task_klass}/).first.squish
```
Comment (etiennebarrie) [inline test/models/maintenance_tasks/task_data_test.rb:59]: Also this test errors when the condition is not respected ("NoMethodError: undefined method `squish' for nil:NilClass"), ideally it would fail:

```suggestion
      assert_includes task_data.code,
        "class #{task_klass} < MaintenanceTasks::Task"
```

File: test/dummy/lib/test_module.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+module TestModule
+  def process(post)
+    sleep(1) unless self.class.fast_task
+
+    post.update!(content: "New (TestModule) content added on #{Time.now.utc}")
```
[the remark this answers is not in the document]
Comment (tyetraskshopify) [reply test/dummy/lib/test_module.rb]: ✅ Great! I wasn't sure if it was critical to differentiate `TestModule#process` for any future usage.

File: test/dummy/app/tasks/maintenance/update_posts_module_prepended_task.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+require_relative "#{Rails.root}/lib/test_module"
```
[the remark this answers is not in the document]
Comment (tyetraskshopify) [reply test/dummy/app/tasks/maintenance/update_posts_module_prepended_task.rb]: ✅ Perfect! I had tried this initially but must have made a typo somewhere because it wasn't working.

File: test/models/maintenance_tasks/task_data_test.rb (hunk 1 above)
Comment (tyetraskshopify) [reply test/models/maintenance_tasks/task_data_test.rb]: ✅ Agreed and updated! I referenced the [existing test](https://github.com/Shopify/maintenance_tasks/blob/main/test/models/maintenance_tasks/task_data_test.rb#L46-L51) of a similar format so I've updated both here.

**Review by etiennebarrie — APPROVED** [review]

**Comment by etiennebarrie** [comment]
Thanks!
